### PR TITLE
feat(ci): build api and web Docker images with Depot

### PR DESCRIPTION
**Pull request by jacobwgillespie**
### What change does this PR introduce? 

This PR switches CI to use [Depot](https://depot.dev) for The `api` and `web` Docker image builds. Depot provides hosted BuildKit build servers, with managed persistent cache, for a 2-3x build speedup. Sending this PR as requested by @scopsy 🙂 

There are a few remaining tasks needed before merge:

- [ ] Create the projects in your Depot organization (feel free to reply here with their IDs and I'll update the PR).
- [ ] Authorize those projects to trust GitHub Actions workflows from this repo.

For the second step, you can add a [OIDC trust relationship](https://depot.dev/docs/cli/authentication#oidc-trust-relationships) in your Depot project settings allowing builds from this repo to use your Depot builders. That will look something like this in the project settings page:

image

If you plan to use Depot for PR builds, you will want to change the Workflows dropdown to `All workflows`.

Feel free to ping if you have any questions or run into any issues!

### Followup work

After this PR merges, there's a few followup things I can submit PRs for:

1. Adding Arm builds - this is mostly a matter of adding `--platform linux/amd64,linux/arm64` to the build commands, but the Dockerfiles need some small modifications to support Arm.
2. Enabling the same changes for the other projects (`embed`, `widget`, and `ws`). I think those could share the same Depot project if you'd like.
3. If you'd like to set up Docker builds for PRs, I can add the code that supports falling back to local `docker build` if a Depot builder isn't available (this happens for PRs from repo forks, since GitHub's security policies block access to OIDC tokens and Actions secrets for those workflows). We have [a GitHub Action](https://github.com/depot/build-push-action) that makes the fallback easy.

### CODE REVIEW

File: .github/workflows/dev-deploy-api.yml [hunk 1]
```diff
@@ -29,6 +29,7 @@ jobs:
       contents: read
       packages: write
       deployments: write
+      id-token: write
```
Comment (jacobwgillespie) [inline .github/workflows/dev-deploy-api.yml:32]: This permission enables access to [GitHub's OIDC token endpoint](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect). This means GitHub will generate a short-lived OIDC token for each build, and Depot uses that token to authenticate. This is in place of a static access token, so you don't have to create one or manage key rotation ✨

File: .github/workflows/dev-deploy-web.yml [hunk 2]
```diff
@@ -107,12 +117,13 @@ jobs:
...
           GH_PASSWORD: ${{ secrets.GH_PACKAGES }}
+          DEPOT_PROJECT_ID: xxxxxx # TODO: add project ID here
         run: |
-          echo $GH_PASSWORD | docker login ghcr.io -u $GH_ACTOR --password-stdin 
-          docker build -t ghcr.io/$REGISTRY_OWNER/$DOCKER_NAME:$IMAGE_TAG -f apps/web/Dockerfile .
-          docker tag ghcr.io/$REGISTRY_OWNER/$DOCKER_NAME:$IMAGE_TAG ghcr.io/$REGISTRY_OWNER/$DOCKER_NAME:dev
-          docker push ghcr.io/$REGISTRY_OWNER/$DOCKER_NAME:dev
-          docker push ghcr.io/$REGISTRY_OWNER/$DOCKER_NAME:$IMAGE_TAG
+          echo $GH_PASSWORD | docker login ghcr.io -u $GH_ACTOR --password-stdin
+          depot build \
```
Comment (djabarovgeorge) [inline .github/workflows/dev-deploy-web.yml:123]:
```suggestion
          depot build --push \
```

Are we missing here a --push flag?

**Review by djabarovgeorge — COMMENTED** [review]
loved the depot project it really improves the workflow speed 🤩

File: .github/workflows/dev-deploy-web.yml (hunk 2 above)
Comment (jacobwgillespie) [reply .github/workflows/dev-deploy-web.yml:123]: We are yes, nice catch 🙂

**Comment by Cliftonz** [comment]
I am good with moving forward with Depot, however, we need to add it to the 4 other containers too.

**Comment by jacobwgillespie** [comment]
@scopsy Once you have created the projects in your Depot org, let me know their IDs and I can update this PR (also let me know if you'd prefer we set that up, happy to help!)

**Comment by scopsy** [comment]
@jacobwgillespie could you add me as a contributor to the fork you have? trying to push the new id's for this branch now

**Comment by jacobwgillespie** [comment]
@scopsy yep added!

**Pull request merged**